File: tools/PhotoLibrarian.ModelBench/review_changes.py
"""Create an incremental, local tag review from saved predictions and human decisions.

Prior result/score pairs are ordered oldest to newest. The latest explicitly
reviewed image/tag wins; disagreement within that review remains pending.
Exact labels are shared across models, never synonyms or bounding-box judgments.
"""

import argparse
import csv
import hashlib
import json
import ntpath
from pathlib import Path
import shutil

import calibrate_scores as cal

EXPORT_FIELDS = cal.PARTIAL_REVIEW_FIELDS
IDENTITY_WARNING = (
    "Older results have no photo content hashes. Reuse assumes a photo at the same "
    "absolute path has not been replaced. Different paths or recorded hashes do not match."
)


def load_run(path, scores=None):
    data = cal.load_inputs(path, scores)
    raw = Path(path).read_bytes()
    cal.require(hashlib.sha256(raw).hexdigest() == data["provenance"]["results"]["sha256"],
                "Predictions changed while loading; retry with a stable saved report.")
    run = json.loads(raw.decode("utf-8-sig"))
    identities, image_paths = {}, {}
    for result in run["Results"]:
        relative = result["RelativeImagePath"]
        image_path = cal.text(result.get("ImagePath"), "ImagePath")
        cal.require(ntpath.isabs(image_path) and bool(ntpath.splitdrive(image_path)[0]),
                    f"{relative}: an absolute Windows ImagePath is required")
        canonical = ntpath.normcase(ntpath.normpath(image_path))
        image_hash = result.get("ImageSha256")
        if image_hash is not None:
            cal.require(isinstance(image_hash, str) and len(image_hash) == 64
                        and all(c in "0123456789abcdefABCDEF" for c in image_hash),
                        f"{relative}: invalid ImageSha256")
            image_hash = image_hash.lower()
        identity = canonical, image_hash
        cal.require(relative not in identities or identities[relative] == identity,
                    f"{relative}: inconsistent photo identity between models")
        identities[relative] = identity
        image_paths[relative] = image_path
    cal.require(len(set(identities.values())) == len(identities),
                "Different image names resolve to the same photo identity")
    return data, identities, image_paths, raw


def thumbnail_path(results_path, image_path):
    stem = ntpath.splitext(ntpath.basename(image_path))[0]
    stem = "".join("_" if c in '<>:"/\\|?*' or ord(c) < 32 else c for c in stem)
    checksum = hashlib.sha256(image_path.encode("utf-8")).hexdigest()[:12].upper()
    return Path(results_path).parent / "thumbnails" / f"{stem}-{checksum}.jpg"


def build_review(results_path, prior_pairs, label_aliases=None):
    cal.require(bool(prior_pairs), "At least one prior results/scores pair is required")
    label_aliases = {} if label_aliases is None else label_aliases
    cal.require(isinstance(label_aliases, dict), "Label aliases must be an object")
    for old, new in label_aliases.items():
        cal.require(isinstance(old, str) and old.strip() == old and bool(old)
                    and isinstance(new, str) and new.strip() == new and bool(new),
                    "Label aliases must contain nonempty, trimmed strings")
        cal.require(new not in label_aliases, "Label aliases cannot contain cycles, self-maps or chains")
    target, identities, image_paths, raw = load_run(results_path)
    history, known = [], {}
    for prior_results, prior_scores in prior_pairs:
        previous, previous_ids, _, _ = load_run(prior_results, prior_scores)
        history.append(previous["provenance"])
        layer = {}
        for review in previous["reviews"]:
            for candidate in review.candidates:
                if candidate.checked is not None:
                    key = previous_ids[review.image], label_aliases.get(candidate.label, candidate.label)
                    layer.setdefault(key, []).append({
                        "model": review.model, "correct": candidate.checked,
                        "source": Path(prior_scores).name,
                    })
        known.update(layer)

    photos, items, rows = [], [], []
    photo_lookup, item_lookup = {}, {}
    for review in target["reviews"]:
        if review.image not in photo_lookup:
            photo_id = len(photos)
            photo_lookup[review.image] = photo_id
            photos.append({"id": photo_id, "image": review.image,
                           "thumbnail": f"thumbnails/{photo_id}.jpg", "items": []})
        photo = photos[photo_lookup[review.image]]
        row = {"image": review.image, "model": review.model, "predictions": []}
        for candidate in review.candidates:
            key = identities[review.image], candidate.label
            if key not in item_lookup:
                evidence = known.get(key, [])
                votes = {entry["correct"] for entry in evidence}
                decision = ("correct" if True in votes else "incorrect") if len(votes) == 1 else None
                item_id = len(items)
                item_lookup[key] = item_id
                photo["items"].append(item_id)
                items.append({
                    "id": item_id, "photo": photo["id"], "label": candidate.label,
                    "initialDecision": decision,
                    "reason": "reused" if decision is not None else "conflict" if votes else "new",
                    "evidence": evidence, "occurrences": [],
                })
            item_id = item_lookup[key]
            items[item_id]["occurrences"].append({
                "model": review.model, "rank": candidate.rank, "score": candidate.score,
                "scoreKind": target["models"][review.model]["scoreKind"],
            })
            row["predictions"].append({"rank": candidate.rank, "item": item_id})
        rows.append(row)

    source = {"target": target["provenance"]["results"], "prior": history}
    if label_aliases:
        source["labelAliases"] = label_aliases
    fingerprint = hashlib.sha256(cal.json_text({"version": 1, **source}).encode("utf-8")).hexdigest()
    data = {
        "schemaVersion": 1, "fingerprint": fingerprint, "identityWarning": IDENTITY_WARNING,
        "provenance": source, "photos": photos, "items": items, "rows": rows,
        "exportFields": EXPORT_FIELDS,
    }
    total = sum(len(row["predictions"]) for row in rows)
    pending = [item for item in items if item["initialDecision"] is None]
    reused_occurrences = sum(len(item["occurrences"]) for item in items if item["initialDecision"] is not None)
    data["summary"] = {
        "photos": len(photos), "modelRows": len(rows), "modelPredictions": total,
        "uniquePhotoTags": len(items), "reusedUniqueDecisions": len(items) - len(pending),
        "reusedModelPredictions": reused_occurrences,
        "pendingUniqueTags": len(pending),
        "newUniqueTags": sum(item["reason"] == "new" for item in pending),
        "conflictingUniqueTags": sum(item["reason"] == "conflict" for item in pending),
        "photosNeedingReview": len({item["photo"] for item in pending}),
        "overallQualityRatingsCopied": 0,
    }
    thumbnails = {
        photo["thumbnail"]: thumbnail_path(results_path, image_paths[photo["image"]])
        for photo in photos
    }
    for thumbnail in thumbnails.values():
        cal.require(thumbnail.is_file(), f"Missing cached thumbnail: {thumbnail}")
    return data, thumbnails, raw


def initial_score_rows(data):
    for row in data["rows"]:
        correct, reviewed = [], []
        for prediction in row["predictions"]:
            decision = data["items"][prediction["item"]]["initialDecision"]
            if decision is not None:
                reviewed.append(str(prediction["rank"]))
                if decision == "correct":
                    correct.append(str(prediction["rank"]))
        yield [row["image"], row["model"], "", ";".join(correct), "", ";".join(reviewed)]


def write_report(output, data, thumbnails, results_bytes):
    output = Path(output)
    cal.require(not output.exists(), f"Output already exists: {output}")
    directory = Path(__file__).parent
    template = (directory / "review_changes.html").read_text(encoding="utf-8")
    cal.require(template.count("@@REVIEW_DATA@@") == 1, "Invalid incremental review template")
    # The payload is inert JSON, but a label must not terminate its script element.
    payload = cal.json_text(data).replace("&", "\\u0026").replace("<", "\\u003c").replace(">", "\\u003e")
    payload = payload.replace("\u2028", "\\u2028").replace("\u2029", "\\u2029")
    page = template.replace("@@REVIEW_DATA@@", payload)
    script = (directory / "review_changes.js").read_text(encoding="utf-8")
    output.mkdir(parents=True, exist_ok=False)
    (output / "thumbnails").mkdir()
    (output / "report.html").write_text(page, encoding="utf-8")
    (output / "review.js").write_text(script, encoding="utf-8")
    (output / "results.json").write_bytes(results_bytes)
    (output / "review-items.json").write_text(cal.json_text(data) + "\n", encoding="utf-8")
    (output / "review-summary.json").write_text(cal.json_text(data["summary"]) + "\n", encoding="utf-8")
    with (output / "human-scores.csv").open("w", newline="", encoding="utf-8-sig") as stream:
        writer = csv.writer(stream)
        writer.writerow(EXPORT_FIELDS)
        writer.writerows(initial_score_rows(data))
    for name, source in thumbnails.items():
        shutil.copyfile(source, output / name)


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--results", required=True, type=Path, help="Target results.json with cached thumbnails")
    parser.add_argument("--prior-results", action="append", required=True, type=Path)
    parser.add_argument("--prior-scores", action="append", required=True, type=Path)
    parser.add_argument("--label-aliases", type=Path,
                        help="Explicit spelling-correction JSON map, or a vocabulary override file with reviewAliases")
    parser.add_argument("--output", required=True, type=Path, help="New report directory; existing reports are preserved")
    args = parser.parse_args(argv)
    try:
        cal.require(len(args.prior_results) == len(args.prior_scores),
                    "Pair each --prior-results with one --prior-scores, oldest to newest")
        cal.require(not args.output.exists(), f"Output already exists: {args.output}")
        aliases = None
        if args.label_aliases:
            aliases = json.loads(args.label_aliases.read_text(encoding="utf-8-sig"))
            if isinstance(aliases, dict) and "reviewAliases" in aliases:
                aliases = aliases["reviewAliases"]
        data, thumbnails, raw = build_review(args.results, list(zip(args.prior_results, args.prior_scores)), aliases)
        write_report(args.output, data, thumbnails, raw)
    except (ValueError, OSError, csv.Error, UnicodeError) as error:
        parser.exit(2, f"Incremental review error: {error}\n")
    print(json.dumps(data["summary"], indent=2))
    print("WARNING: " + IDENTITY_WARNING)
    print(f"Review: {args.output / 'report.html'}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/PhotoLibrarian.ModelBench/review_changes.js
"use strict";
(() => {
  const data = JSON.parse(document.getElementById("review-data").textContent);
  const byId = id => document.getElementById(id);
  const decisions = data.items.map(item => item.initialDecision);
  const storageKey = "PhotoLibrarian.ModelBench.Changes:" + data.fingerprint;
  let currentPhoto = null;
  let shownItems = [];
  let overrides = {};
  let storageUsable = true;

  function error(message) {
    byId("error").textContent = message;
    byId("error").hidden = false;
  }
  try {
    const saved = localStorage.getItem(storageKey);
    if (saved !== null) {
      const parsed = JSON.parse(saved);
      if (!parsed || parsed.fingerprint !== data.fingerprint || !parsed.decisions ||
          typeof parsed.decisions !== "object" || Array.isArray(parsed.decisions)) {
        throw new Error("Unrecognized saved review format.");
      }
      for (const [key, value] of Object.entries(parsed.decisions)) {
        if (!/^(0|[1-9]\d*)$/.test(key) || !data.items[Number(key)] ||
            ![null, "correct", "incorrect"].includes(value)) {
          throw new Error("Invalid saved tag decision.");
        }
      }
      overrides = parsed.decisions;
      for (const [key, value] of Object.entries(overrides)) decisions[Number(key)] = value;
    }
  } catch (problem) {
    storageUsable = false;
    error("Saved browser review could not be loaded. Previous CSV decisions remain available. " +
      "Browser saving is disabled to avoid overwriting saved work; export your changes. " + problem.message);
  }

  function save() {
    if (!storageUsable) return;
    try {
      localStorage.setItem(storageKey, JSON.stringify({fingerprint: data.fingerprint, decisions: overrides}));
    } catch (problem) {
      error("Your edits are retained in this page but could not be saved in the browser. Export before closing. " + problem.message);
    }
  }
  const pending = photo => photo.items.filter(id => decisions[id] === null);
  const showReviewed = () => byId("show-reviewed").checked;

  function updateProgress() {
    const remaining = data.items.filter(item => decisions[item.id] === null).length;
    const photosLeft = data.photos.filter(photo => pending(photo).length > 0).length;
    byId("progress").textContent =
      `${remaining} tags on ${photosLeft} photos left to review. ` +
      `${data.summary.reusedUniqueDecisions} prior decisions reused; ${data.summary.conflictingUniqueTags} prior conflicts flagged.`;
    byId("complete").hidden = remaining !== 0;
    const select = byId("photo-select");
    select.replaceChildren();
    const visible = data.photos.filter(photo => showReviewed() || pending(photo).length > 0 || photo.id === currentPhoto);
    for (const photo of visible) {
      const option = document.createElement("option");
      option.value = String(photo.id);
      option.textContent = `${photo.image} (${pending(photo).length} pending)`;
      select.append(option);
    }
    if (currentPhoto !== null) select.value = String(currentPhoto);
    select.disabled = visible.length === 0;
    byId("previous").disabled = byId("next").disabled = visible.length <= 1;
    if (currentPhoto !== null) {
      const photo = data.photos[currentPhoto];
      byId("photo-progress").textContent = `${pending(photo).length} pending; ${photo.items.length} distinct suggested tags on this photo.`;
      byId("reject-remaining").disabled = pending(photo).length === 0;
    }
  }

  function applyDecision(id, value) {
    decisions[id] = value;
    overrides[String(id)] = value;
    save();
  }

  function renderTags() {
    const container = byId("tags");
    container.replaceChildren();
    for (const id of shownItems) {
      const item = data.items[id];
      const card = document.createElement("article");
      card.className = "tag";
      card.dataset.item = String(id);
      const heading = document.createElement("h3");
      heading.textContent = item.label;
      card.append(heading);
      const status = document.createElement("p");
      status.className = "decision";
      status.textContent = decisions[id] === null
        ? (item.reason === "conflict" ? "Pending: prior reviews disagree" : "Pending: not previously reviewed")
        : decisions[id] === "correct" ? "Correct" : "Rejected";
      card.append(status);
      const occurrences = document.createElement("p");
      occurrences.className = "muted";
      occurrences.textContent = item.occurrences.map(p =>
        `${p.model} #${p.rank}: ${p.scoreKind === "binary" ? "detected" : `${Number(p.score).toPrecision(4)} ${p.scoreKind}`}`).join(" | ");
      card.append(occurrences);
      if (item.reason === "conflict") {
        const evidence = document.createElement("p");
        evidence.className = "muted";
        evidence.textContent = item.evidence.map(p =>
          `${p.model}: ${p.correct ? "correct" : "rejected"} (${p.source})`).join("; ");
        card.append(evidence);
      }
      const actions = document.createElement("div");
      actions.className = "actions";
      for (const [label, value] of [["Correct", "correct"], ["Reject", "incorrect"], ["Not sure", null]]) {
        const button = document.createElement("button");
        button.type = "button";
        button.textContent = label;
        button.setAttribute("aria-label", `${label}: ${item.label}`);
        button.setAttribute("aria-pressed", String(decisions[id] === value));
        button.addEventListener("click", () => {
          applyDecision(id, value);
          renderTags();
          updateProgress();
          const replacement = byId("tags").querySelector(`[data-item="${id}"]`);
          replacement.querySelectorAll("button")[value === "correct" ? 0 : value === "incorrect" ? 1 : 2].focus();
        });
        actions.append(button);
      }
      card.append(actions);
      container.append(card);
    }
  }

  function renderPhoto(id) {
    currentPhoto = id;
    byId("workspace").hidden = id === null;
    if (id !== null) {
      const photo = data.photos[id];
      byId("photo-name").textContent = photo.image;
      byId("photo").src = photo.thumbnail;
      byId("photo").alt = photo.image;
      shownItems = photo.items.filter(item => showReviewed() || decisions[item] === null);
      renderTags();
    }
    updateProgress();
  }

  function navigate(direction) {
    const options = [...byId("photo-select").options].map(option => Number(option.value));
    if (options.length === 0) return;
    const index = options.indexOf(currentPhoto);
    renderPhoto(options[(index + direction + options.length) % options.length]);
  }

  byId("identity-warning").textContent = data.identityWarning;
  byId("previous").addEventListener("click", () => navigate(-1));
  byId("next").addEventListener("click", () => navigate(1));
  byId("photo-select").addEventListener("change", event => renderPhoto(Number(event.target.value)));
  byId("show-reviewed").addEventListener("change", () => {
    const current = currentPhoto === null ? null : data.photos[currentPhoto];
    const photo = current && (showReviewed() || pending(current).length > 0)
      ? current : data.photos.find(p => showReviewed() || pending(p).length > 0);
    renderPhoto(photo ? photo.id : null);
  });
  byId("reject-remaining").addEventListener("click", () => {
    if (currentPhoto === null) return;
    const ids = pending(data.photos[currentPhoto]);
    if (ids.length === 0 || !confirm(`Reject all ${ids.length} still-pending tags on this photo?`)) return;
    for (const id of ids) applyDecision(id, "incorrect");
    renderTags();
    updateProgress();
  });
  byId("export").addEventListener("click", () => {
    save();
    const rows = [data.exportFields];
    for (const row of data.rows) {
      const reviewed = row.predictions.filter(p => decisions[p.item] !== null).map(p => p.rank);
      const correct = row.predictions.filter(p => decisions[p.item] === "correct").map(p => p.rank);
      rows.push([row.image, row.model, "", correct.join(";"), "", reviewed.join(";")]);
    }
    const csv = value => '"' + String(value ?? "").replaceAll('"', '""') + '"';
    const content = "\uFEFF" + rows.map(row => row.map(csv).join(",")).join("\r\n");
    const url = URL.createObjectURL(new Blob([content], {type: "text/csv;charset=utf-8"}));
    const link = document.createElement("a");
    link.href = url;
    link.download = "human-scores-delta.csv";
    document.body.append(link);
    link.click();
    link.remove();
    setTimeout(() => URL.revokeObjectURL(url), 1000);
  });
  const first = data.photos.find(photo => pending(photo).length > 0);
  renderPhoto(first ? first.id : null);
})();
